source/extensions/common/wasm/foreign.cc

```diff
@@ -126,6 +126,14 @@ RegisterForeignFunction registerSetEnvoyFilterStateForeignFunction(
       return WasmResult::BadArgument;
     });
 
+RegisterForeignFunction registerClearRouteCacheForeignFunction(
+    "clear_route_cache",
+    [](WasmBase&, std::string_view, const std::function<void*(size_t size)>&) -> WasmResult {
+      auto context = static_cast<Context*>(proxy_wasm::current_context_);
+      context->clearRouteCache();
```

Comment:
nit: Is this safe to call at any point in execution? This is untrusted code calling out, we might need to harden it if it's not already done.

Reply:
I remember I did check in the clearRouteCache self. Will re-check it again.

Reply:
The clearRouteCache will be forbided after the response headers is processed. But I agree we should harden it continuously to ensure the interface that's provided to extension is robust enough.

For example, only allow the head filter to clear the route cache. I will do this with another PR.

Reply:
I think from the separate discussion in proxy-wasm-spec, what we need is a control knob on the filter xDS with several options:
1. Clear route cache on every header operation (past default).
2. No clear route cache at all (future default).
3. Permit clear route cache external function (to disallow untrusted extensions from causing harm by clearing routes).
4. Automatic clear route after callback (to mimic ext_proc, not needed for your use case but it will be easier to handle dynamic metadata routing, for example).

We can split this into several booleans as well (e.g. 3 - could be a separate knob).
@PiotrSikora - does that look right to you?

Reply:
Now, this PR removed the automatic route refresh and provide a manual way. This should be good enough as a default way to handle the route cache.

New options (3 and 4) could be treated as independent requirement of Envoy-self and be implemented in the future.

I think a basic rule is `listening the feedbacks from acutal practices and responding them quickly`, so, if some users have actual related requirements, I am happy to do it. :)

Reply:
@kyessenov yeah, that sounds good, although I'm not sure I understand the difference between 1 and 4.

Reply:
> @kyessenov yeah, that sounds good, although I'm not sure I understand the difference between 1 and 4.

The difference is that in which case we need additional xDS control. In the new implementation, the automical refresh should be enabled by an explict option.

Reply:
@PiotrSikora The difference is that 1) clears route only on header operations but 4) should clear routes on any dynamic metadata or other non-HTTP attribute update